src/lib/ctable.lua

```diff
@@ -0,0 +1,466 @@
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local C = ffi.C
+local S = require("syscall")
+local bit = require("bit")
+local bxor, bnot = bit.bxor, bit.bnot
+local tobit, lshift, rshift = bit.tobit, bit.lshift, bit.rshift
+local max, floor, ceil = math.max, math.floor, math.ceil
+
+CTable = {}
+
+local HASH_MAX = 0xFFFFFFFF
+local uint16_ptr_t = ffi.typeof('uint16_t*')
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+local uint64_ptr_t = ffi.typeof('uint64_t*')
+
+local function make_entry_type(key_type, value_type)
+   return ffi.typeof([[struct {
+         uint32_t hash;
+         $ key;
+         $ value;
+      } __attribute__((packed))]],
+      key_type,
+      value_type)
+end
+
+local function make_entries_type(entry_type)
+   return ffi.typeof('$[?]', entry_type)
+end
+
+-- hash := [0,HASH_MAX); scale := size/HASH_MAX
+local function hash_to_index(hash, scale)
+   return floor(hash*scale + 0.5)
+end
+
+local function make_equal_fn(key_type)
+   local size = ffi.sizeof(key_type)
+   local cast = ffi.cast
+   if tonumber(ffi.new(key_type)) then
+      return function (a, b)
+         return a == b
+      end
+   elseif size == 2 then
+      return function (a, b)
+         return cast(uint16_ptr_t, a)[0] == cast(uint16_ptr_t, b)[0]
+      end
+   elseif size == 4 then
+      return function (a, b)
+         return cast(uint32_ptr_t, a)[0] == cast(uint32_ptr_t, b)[0]
+      end
+   elseif size == 6 then
+      return function (a, b)
+         return (cast(uint32_ptr_t, a)[0] == cast(uint32_ptr_t, b)[0] and
+                 cast(uint16_ptr_t, a)[2] == cast(uint16_ptr_t, b)[2])
+      end
+   elseif size == 8 then
+      return function (a, b)
+         return cast(uint64_ptr_t, a)[0] == cast(uint64_ptr_t, b)[0]
+      end
+   else
+      return function (a, b)
+         return C.memcmp(a, b, size) == 0
+      end
+   end
+end
+
+local function set(...)
+   local ret = {}
+   for k, v in pairs({...}) do ret[v] = true end
+   return ret
+end
+
+local function parse_params(params, required, optional)
+   local ret = {}
+   for k, _ in pairs(required) do
+      if params[k] == nil then error('missing required option ' .. k) end
+   end
+   for k, v in pairs(params) do
+      if not required[k] and optional[k] == nil then
+         error('unrecognized option ' .. k)
+      end
+      ret[k] = v
+   end
+   for k, v in pairs(optional) do
+      if ret[k] == nil then ret[k] = v end
+   end
+   return ret
+end
+
+-- FIXME: For now the value_type option is required, but in the future
+-- we should allow for a nil value type to create a set instead of a
+-- map.
+local required_params = set('key_type', 'value_type', 'hash_fn')
+local optional_params = {
+   initial_size = 8,
+   max_occupancy_rate = 0.9,
+   min_occupancy_rate = 0.0
+}
+
+function new(params)
+   local ctab = {}   
+   local params = parse_params(params, required_params, optional_params)
```

Comment:
I can relate to the last point, might be worth to spin this out into `core.lib` in the future then. We already have some common code for app configurations which then should be unified with this.